feat/3324-adding-iceberg-catalogs-compatibilities

lfagliano opened the pull request:
### Description

This PR removes the hardcoded in-memory SQLite catalog limitation and enables DLT to work with all PyIceberg-supported catalog types including REST catalogs, through leveraging the `load_catalog` functionality. The actual execution and operation of Iceberg format remains the same, we just load the catalog and pass them along to the pipeline.  

Previously, the Iceberg destination hardcoded catalog creation to use `sqlite:///:memory:`, limiting Iceberg support:

```python
catalog = get_sql_catalog(
    catalog_name or "default", 
    "sqlite:///:memory:",  # No other options!
    self.config.credentials
)
```

Leverage PyIceberg's built-in `load_catalog()` function to support all catalog types through standard configuration methods. The implementation provides a priority chain that tries multiple configuration sources and falls back gracefully.

In essence, this allows the user to add a `.pyiceberg.yaml` file to reference to their catalogs (which is in line with pyiceberg), similar to this: 
```yaml
# ~/.pyiceberg.yaml
catalog:
  production:
    type: rest
    uri: https://localhost:8181/catalog
    warehouse: analytics
```

The changes I made do not require the user to do much more than that, and they can continue using Iceberg as they were. 


### Related Issues

- Fixes #3324 

### Additional Context

To test it I run this for some time, monkey patching the core package. I then also added my contributions, built the package and tried it again through two of my prod pipelines. 

As for the tests, I run them locally, but I came across some what I think are broken tests -- because they had absolutely nothing to do with my changes, but I am gonna be checking them more in detail tomorrow.

rudolfix commented on dlt/common/libs/pyiceberg.py:
```diff
@@ -142,6 +144,273 @@ def get_sql_catalog(
+        # Search in standard locations
+        search_paths.extend(
+            [
+                Path.cwd() / ".pyiceberg.yaml",
```
please take a look at `dlt.current.run_context()` and use
```
def run_dir(self) -> str:
        """Defines the context working directory, defaults to cwd()"""
```

rudolfix commented on dlt/common/libs/pyiceberg.py:
```diff
@@ -142,6 +144,273 @@ def get_sql_catalog(
+
+    Args:
+        catalog_name: Name of the catalog (default: "default")
+        catalog_type: Type of catalog ('sql' or 'rest') - used for fallback
```
this is not passed and IMO should

rudolfix commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -1000,16 +1000,47 @@ def get_open_table_catalog(self, table_format: TTableFormat, catalog_name: str =
+        try:
+            catalog.create_namespace(self.dataset_name)
+            logger.info(f"Created Iceberg namespace: {self.dataset_name}")
+        except Exception as e:
```
please catch more specific exception (is there exception for entity to already exist?)

lfagliano commented on tests/common/libs/pytest.ini:
```diff
@@ -0,0 +1,27 @@
+[pytest]
```
removed! this was a slip from my local dev setup

lfagliano commented:
@rudolfix  thanks for the review! I addressed your comments, let me know if you would like me to change anything else. 

About the tests, only today I saw your comment and I had already started over the weekend with the integration tests trying to do a similar setup to `pyiceberg`, and I was about to commit them. I tried to re-work them a bit. Basically I am adding a `docker-compose` file that can be used to set up the whole testing environment with the two catalogs (rest & postgres), as well as a `init-iceberg-catalogs.py` which sets up the catalogs for the tests. Originally I thought that would be what we would put in CI, but for now I added `skip` to the REST and Postgres fixtures, so that they are essentially optional. I have been using them for the last couple of days and found them useful. Let me know what you think!

lfagliano commented:
@rudolfix I was referring to the current integration tests in the PR, which they should already be skipped if they dont detect any catalog present (btw, I moved away from lakekeeper for the tests, instead I am using the apache testing [image](https://hub.docker.com/r/apache/iceberg-rest-fixture) )

But then, I think I'm happy with moving forward with what we have already :D 🚀 Is there anything else you need me to do? I just corrected the lint errors
